File: backend/schemas/reviews.py
"""
Review workflow schemas for request/response models.
Provides Pydantic models for internal review workflow API endpoints.
"""
from datetime import datetime
from enum import Enum
from typing import Any, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field, computed_field, field_validator, model_validator

from backend.schemas.common import PaginatedResponse, PaginationInfo
from backend.schemas.jsonb_types import (
    WorkflowStageDict,
    TeamMemberPermissionsDict,
    ReviewActionMetadataDict,
)


class ReviewStatus(str, Enum):
    """Status values for application reviews."""
    PENDING = "pending"
    IN_REVIEW = "in_review"
    APPROVED = "approved"
    REJECTED = "rejected"
    ESCALATED = "escalated"


class ReviewAction(str, Enum):
    """Actions that can be taken during a review."""
    APPROVED = "approved"
    REJECTED = "rejected"
    RETURNED = "returned"
    COMMENTED = "commented"


class TeamMemberRole(str, Enum):
    """Roles for application team members."""
    PI = "pi"
    CO_I = "co_i"
    GRANT_WRITER = "grant_writer"
    REVIEWER = "reviewer"
    ADMIN = "admin"


# Status display configuration for UI
REVIEW_STATUS_CONFIG = {
    ReviewStatus.PENDING: {"label": "Pending", "color": "gray", "order": 0},
    ReviewStatus.IN_REVIEW: {"label": "In Review", "color": "blue", "order": 1},
    ReviewStatus.APPROVED: {"label": "Approved", "color": "green", "order": 2},
    ReviewStatus.REJECTED: {"label": "Rejected", "color": "red", "order": 3},
    ReviewStatus.ESCALATED: {"label": "Escalated", "color": "orange", "order": 4},
}

# Role display configuration for UI
ROLE_CONFIG = {
    TeamMemberRole.PI: {"label": "Principal Investigator", "short": "PI"},
    TeamMemberRole.CO_I: {"label": "Co-Investigator", "short": "Co-I"},
    TeamMemberRole.GRANT_WRITER: {"label": "Grant Writer", "short": "Writer"},
    TeamMemberRole.REVIEWER: {"label": "Reviewer", "short": "Rev"},
    TeamMemberRole.ADMIN: {"label": "Administrator", "short": "Admin"},
}


# ============================================================================
# Workflow Stage Schemas
# ============================================================================

class WorkflowStage(BaseModel):
    """Schema for a workflow stage configuration."""
    order: int = Field(..., ge=0, description="Stage order (0-indexed)")
    name: str = Field(..., min_length=1, max_length=100, description="Stage name")
    required_role: Optional[TeamMemberRole] = Field(None, description="Required role to approve this stage")
    sla_hours: Optional[int] = Field(None, ge=1, description="SLA time limit in hours")
    auto_escalate: bool = Field(False, description="Whether to auto-escalate when SLA is exceeded")
    description: Optional[str] = Field(None, max_length=500, description="Stage description")


class WorkflowStageResponse(WorkflowStage):
    """Response schema for workflow stage with computed fields."""

    @computed_field
    @property
    def role_config(self) -> Optional[dict]:
        """Get role display configuration."""
        if self.required_role:
            return ROLE_CONFIG.get(self.required_role, {"label": self.required_role.value, "short": self.required_role.value})
        return None


# ============================================================================
# Review Workflow Schemas
# ============================================================================

class ReviewWorkflowCreate(BaseModel):
    """Schema for creating a new review workflow."""
    name: str = Field(..., min_length=1, max_length=255, description="Workflow name")
    description: Optional[str] = Field(None, max_length=2000, description="Workflow description")
    stages: List[WorkflowStage] = Field(..., min_length=1, description="Workflow stages")
    is_default: bool = Field(False, description="Set as default workflow")

    @field_validator("stages")
    @classmethod
    def validate_stages_structure(cls, v: List[WorkflowStage]) -> List[WorkflowStage]:
        """Validate that stages conform to WorkflowStageDict structure and are properly ordered."""
        if not v:
            raise ValueError("At least one stage is required")

        # Validate stage orders are sequential starting from 0
        orders = sorted(stage.order for stage in v)
        expected = list(range(len(v)))
        if orders != expected:
            raise ValueError(f"Stage orders must be sequential starting from 0. Got: {orders}")

        # Validate stage names are unique
        names = [stage.name for stage in v]
        if len(names) != len(set(names)):
            raise ValueError("Stage names must be unique")

        return v


class ReviewWorkflowUpdate(BaseModel):
    """Schema for updating a review workflow."""
    name: Optional[str] = Field(None, min_length=1, max_length=255, description="Workflow name")
    description: Optional[str] = Field(None, max_length=2000, description="Workflow description")
    stages: Optional[List[WorkflowStage]] = Field(None, min_length=1, description="Workflow stages")
    is_default: Optional[bool] = Field(None, description="Set as default workflow")
    is_active: Optional[bool] = Field(None, description="Whether workflow is active")

    @field_validator("stages")
    @classmethod
    def validate_stages_structure(cls, v: Optional[List[WorkflowStage]]) -> Optional[List[WorkflowStage]]:
        """Validate that stages conform to WorkflowStageDict structure if provided."""
        if v is None:
            return v

        if not v:
            raise ValueError("At least one stage is required when updating stages")

        # Validate stage orders are sequential starting from 0
        orders = sorted(stage.order for stage in v)
        expected = list(range(len(v)))
        if orders != expected:
            raise ValueError(f"Stage orders must be sequential starting from 0. Got: {orders}")

        # Validate stage names are unique
        names = [stage.name for stage in v]
        if len(names) != len(set(names)):
            raise ValueError("Stage names must be unique")

        return v


class ReviewWorkflowResponse(BaseModel):
    """Schema for review workflow response."""
    id: UUID = Field(..., description="Workflow ID")
    user_id: UUID = Field(..., description="Owner user ID")
    name: str = Field(..., description="Workflow name")
    description: Optional[str] = Field(None, description="Workflow description")
    stages: List[WorkflowStageResponse] = Field(..., description="Workflow stages")
    is_default: bool = Field(..., description="Whether this is the default workflow")
    is_active: bool = Field(..., description="Whether this workflow is active")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")

    @computed_field
    @property
    def stage_count(self) -> int:
        """Number of stages in the workflow."""
        return len(self.stages)

    class Config:
        from_attributes = True


class ReviewWorkflowList(BaseModel):
    """Schema for list of review workflows (standard paginated format)."""
    data: List[ReviewWorkflowResponse] = Field(..., description="List of workflows")
    pagination: PaginationInfo = Field(..., description="Pagination metadata")

    # Backward compatibility aliases
    @property
    def items(self) -> List[ReviewWorkflowResponse]:
        """Backward compatibility: alias for 'data'."""
        return self.data

    @property
    def total(self) -> int:
        """Backward compatibility: alias for 'pagination.total'."""
        return self.pagination.total


# ============================================================================
# Application Review Schemas
# ============================================================================

class StartReviewRequest(BaseModel):
    """Schema for starting a review process."""
    workflow_id: Optional[UUID] = Field(None, description="Workflow ID (uses default if not provided)")


class ReviewActionRequest(BaseModel):
    """Schema for submitting a review action."""
    action: ReviewAction = Field(..., description="Action to take")
    comments: Optional[str] = Field(None, max_length=5000, description="Comments/feedback")
    metadata: Optional[dict[str, Any]] = Field(None, description="Additional action metadata")


class ReviewStageActionResponse(BaseModel):
    """Schema for a review stage action response."""
    id: UUID = Field(..., description="Action ID")
    review_id: UUID = Field(..., description="Review ID")
    stage_order: int = Field(..., description="Stage index when action was taken")
    stage_name: str = Field(..., description="Stage name when action was taken")
    reviewer_id: Optional[UUID] = Field(None, description="Reviewer user ID")
    reviewer_name: Optional[str] = Field(None, description="Reviewer name")
    reviewer_email: Optional[str] = Field(None, description="Reviewer email")
    action: ReviewAction = Field(..., description="Action taken")
    comments: Optional[str] = Field(None, description="Comments/feedback")
    metadata: Optional[dict[str, Any]] = Field(None, description="Additional action metadata")
    acted_at: datetime = Field(..., description="When the action was taken")

    @computed_field
    @property
    def action_label(self) -> str:
        """Human-readable action label."""
        labels = {
            ReviewAction.APPROVED: "Approved",
            ReviewAction.REJECTED: "Rejected",
            ReviewAction.RETURNED: "Returned for Revision",
            ReviewAction.COMMENTED: "Added Comment",
        }
        return labels.get(self.action, self.action.value)

    class Config:
        from_attributes = True


class ApplicationReviewResponse(BaseModel):
    """Schema for application review response."""
    id: UUID = Field(..., description="Review ID")
    kanban_card_id: UUID = Field(..., description="Application ID")
    workflow_id: Optional[UUID] = Field(None, description="Workflow ID")
    workflow_name: Optional[str] = Field(None, description="Workflow name")
    current_stage: int = Field(..., description="Current stage index")
    current_stage_name: Optional[str] = Field(None, description="Current stage name")
    total_stages: int = Field(0, description="Total number of stages")
    status: ReviewStatus = Field(..., description="Review status")
    started_by: Optional[UUID] = Field(None, description="User who started the review")
    started_by_name: Optional[str] = Field(None, description="Name of user who started the review")
    started_at: datetime = Field(..., description="When the review was started")
    completed_at: Optional[datetime] = Field(None, description="When the review was completed")
    stage_started_at: datetime = Field(..., description="When the current stage started")
    escalation_sent: bool = Field(..., description="Whether escalation was sent for current stage")
    sla_hours: Optional[int] = Field(None, description="SLA hours for current stage")
    actions: List[ReviewStageActionResponse] = Field(default_factory=list, description="Review actions history")

    @computed_field
    @property
    def status_config(self) -> dict:
        """Get status display configuration."""
        return REVIEW_STATUS_CONFIG.get(self.status, {"label": self.status.value, "color": "gray", "order": 99})

    @computed_field
    @property
    def progress_percent(self) -> float:
        """Calculate review progress percentage."""
        if self.total_stages == 0:
            return 0.0
        if self.status in [ReviewStatus.APPROVED, ReviewStatus.REJECTED]:
            return 100.0
        return round((self.current_stage / self.total_stages) * 100, 1)

    class Config:
        from_attributes = True


class ReviewHistoryResponse(BaseModel):
    """Schema for review history (all actions)."""
    review: ApplicationReviewResponse = Field(..., description="Review details")
    actions: List[ReviewStageActionResponse] = Field(..., description="All actions taken")


# ============================================================================
# Team Member Schemas
# ============================================================================

class TeamMemberPermissions(BaseModel):
    """Schema for team member permissions."""
    can_edit: bool = Field(True, description="Can edit application content")
    can_approve: bool = Field(False, description="Can approve review stages")
    can_submit: bool = Field(False, description="Can submit the application")
    sections: Optional[List[str]] = Field(None, description="Specific sections they can access")


class AddTeamMemberRequest(BaseModel):
    """Schema for adding a team member to an application."""
    user_id: Optional[UUID] = Field(None, description="User ID (if known)")
    email: Optional[str] = Field(None, description="User email (to invite)")
    role: TeamMemberRole = Field(..., description="Role in the application")
    permissions: Optional[TeamMemberPermissions] = Field(None, description="Custom permissions")


class UpdateTeamMemberRequest(BaseModel):
    """Schema for updating a team member."""
    role: Optional[TeamMemberRole] = Field(None, description="New role")
    permissions: Optional[TeamMemberPermissions] = Field(None, description="Updated permissions")


class TeamMemberResponse(BaseModel):
    """Schema for team member response."""
    id: UUID = Field(..., description="Team member record ID")
    kanban_card_id: UUID = Field(..., description="Application ID")
    user_id: UUID = Field(..., description="User ID")
    user_name: Optional[str] = Field(None, description="User name")
    user_email: Optional[str] = Field(None, description="User email")
    role: TeamMemberRole = Field(..., description="Role in the application")
    permissions: Optional[TeamMemberPermissions] = Field(None, description="Custom permissions")
    added_by: Optional[UUID] = Field(None, description="User who added this member")
    added_at: datetime = Field(..., description="When the member was added")

    @computed_field
    @property
    def role_config(self) -> dict:
        """Get role display configuration."""
        return ROLE_CONFIG.get(self.role, {"label": self.role.value, "short": self.role.value})

    @computed_field
    @property
    def default_permissions(self) -> TeamMemberPermissions:
        """Get default permissions based on role."""
        defaults = {
            TeamMemberRole.PI: TeamMemberPermissions(can_edit=True, can_approve=True, can_submit=True),
            TeamMemberRole.CO_I: TeamMemberPermissions(can_edit=True, can_approve=True, can_submit=False),
            TeamMemberRole.GRANT_WRITER: TeamMemberPermissions(can_edit=True, can_approve=False, can_submit=False),
            TeamMemberRole.REVIEWER: TeamMemberPermissions(can_edit=False, can_approve=True, can_submit=False),
            TeamMemberRole.ADMIN: TeamMemberPermissions(can_edit=True, can_approve=True, can_submit=True),
        }
        return defaults.get(self.role, TeamMemberPermissions())

    class Config:
        from_attributes = True


class TeamMemberList(BaseModel):
    """Schema for list of team members (standard paginated format)."""
    data: List[TeamMemberResponse] = Field(..., description="List of team members")
    pagination: PaginationInfo = Field(..., description="Pagination metadata")

    # Backward compatibility aliases
    @property
    def items(self) -> List[TeamMemberResponse]:
        """Backward compatibility: alias for 'data'."""
        return self.data

    @property
    def total(self) -> int:
        """Backward compatibility: alias for 'pagination.total'."""
        return self.pagination.total


# ============================================================================
# Default Workflow Templates
# ============================================================================

DEFAULT_WORKFLOWS = {
    "standard": {
        "name": "Standard Review",
        "description": "Standard internal review workflow for grant applications",
        "stages": [
            {"order": 0, "name": "Draft Review", "required_role": "grant_writer", "sla_hours": 48, "auto_escalate": False},
            {"order": 1, "name": "PI Review", "required_role": "pi", "sla_hours": 72, "auto_escalate": True},
            {"order": 2, "name": "Department Approval", "required_role": "admin", "sla_hours": 96, "auto_escalate": True},
            {"order": 3, "name": "Sponsored Programs", "required_role": "admin", "sla_hours": 120, "auto_escalate": True},
        ],
    },
    "quick": {
        "name": "Quick Review",
        "description": "Expedited review for smaller grants or renewals",
        "stages": [
            {"order": 0, "name": "PI Review", "required_role": "pi", "sla_hours": 24, "auto_escalate": False},
            {"order": 1, "name": "Admin Approval", "required_role": "admin", "sla_hours": 48, "auto_escalate": False},
        ],
    },
    "comprehensive": {
        "name": "Comprehensive Review",
        "description": "Full review cycle for major grants",
        "stages": [
            {"order": 0, "name": "Draft Review", "required_role": "grant_writer", "sla_hours": 48, "auto_escalate": False},
            {"order": 1, "name": "Co-I Review", "required_role": "co_i", "sla_hours": 72, "auto_escalate": False},
            {"order": 2, "name": "PI Review", "required_role": "pi", "sla_hours": 72, "auto_escalate": True},
            {"order": 3, "name": "Department Chair", "required_role": "admin", "sla_hours": 96, "auto_escalate": True},
            {"order": 4, "name": "College Review", "required_role": "admin", "sla_hours": 96, "auto_escalate": True},
            {"order": 5, "name": "Sponsored Programs", "required_role": "admin", "sla_hours": 120, "auto_escalate": True},
        ],
    },
}


class DefaultWorkflowResponse(BaseModel):
    """Schema for default workflow template."""
    key: str = Field(..., description="Template key")
    name: str = Field(..., description="Template name")
    description: str = Field(..., description="Template description")
    stages: List[WorkflowStage] = Field(..., description="Template stages")
    stage_count: int = Field(..., description="Number of stages")

File: frontend/src/types/reviews.ts
import type { User } from './index';

// Role types for team members in a review workflow
export type TeamMemberRole = 'pi' | 'co_i' | 'grant_writer' | 'reviewer' | 'admin';

// Review status types
export type ReviewStatus = 'pending' | 'in_review' | 'approved' | 'rejected' | 'escalated';

// Review action types
export type ReviewAction = 'approved' | 'rejected' | 'returned' | 'commented';

// Review stage configuration
export interface ReviewStage {
  order: number;
  name: string;
  required_role: TeamMemberRole;
  sla_hours: number;
  auto_escalate: boolean;
}

// Review workflow template
export interface ReviewWorkflow {
  id: string;
  name: string;
  stages: ReviewStage[];
  is_default: boolean;
  created_at: string;
}

// Application-specific review instance
export interface ApplicationReview {
  id: string;
  kanban_card_id: string;
  workflow_id: string;
  workflow?: ReviewWorkflow;
  current_stage: number;
  status: ReviewStatus;
  started_at: string;
  completed_at?: string;
  /**
   * Computed fields - calculated by the backend based on workflow and current state.
   * These should NOT be computed on the frontend; they are returned from the API response.
   */
  /** Current stage name derived from workflow.stages[current_stage] on backend */
  current_stage_name?: string;
  /** SLA deadline calculated from started_at + stage.sla_hours on backend */
  sla_deadline?: string;
  /** Whether the review is past its SLA deadline, calculated on backend */
  is_overdue?: boolean;
  /** Progress percentage through the workflow stages, calculated on backend */
  progress_percent?: number;
}

// Individual review action record
export interface ReviewStageAction {
  id: string;
  review_id: string;
  stage_order: number;
  stage_name?: string;
  reviewer_id: string;
  reviewer?: User;
  action: ReviewAction;
  comments?: string;
  acted_at: string;
}

// Team member permissions
export interface TeamMemberPermissions {
  can_edit: boolean;
  can_approve: boolean;
  can_submit: boolean;
  sections?: string[];
}

// Team member assigned to an application
export interface ApplicationTeamMember {
  id: string;
  kanban_card_id: string;
  user_id: string;
  user?: User;
  role: TeamMemberRole;
  permissions: TeamMemberPermissions;
  added_at: string;
}

// API request types
export interface StartReviewRequest {
  workflow_id?: string;
}

export interface SubmitReviewActionRequest {
  action: ReviewAction;
  comments?: string;
}

export interface AddTeamMemberRequest {
  user_id?: string;
  email?: string;
  role: TeamMemberRole;
  permissions?: Partial<TeamMemberPermissions>;
}

export interface UpdateTeamMemberRequest {
  role?: TeamMemberRole;
  permissions?: Partial<TeamMemberPermissions>;
}

// API response types
export interface ReviewWorkflowListResponse {
  workflows: ReviewWorkflow[];
  total: number;
}

export interface ReviewHistoryResponse {
  actions: ReviewStageAction[];
  total: number;
}

export interface TeamMembersResponse {
  members: ApplicationTeamMember[];
  total: number;
}

// UI configuration for roles
export interface RoleConfig {
  key: TeamMemberRole;
  label: string;
  color: string;
  bgColor: string;
  description: string;
}

export const ROLE_CONFIGS: Record<TeamMemberRole, RoleConfig> = {
  pi: {
    key: 'pi',
    label: 'Principal Investigator',
    color: 'text-purple-700',
    bgColor: 'bg-purple-100',
    description: 'Lead researcher and primary contact',
  },
  co_i: {
    key: 'co_i',
    label: 'Co-Investigator',
    color: 'text-blue-700',
    bgColor: 'bg-blue-100',
    description: 'Supporting researcher with shared responsibility',
  },
  grant_writer: {
    key: 'grant_writer',
    label: 'Grant Writer',
    color: 'text-emerald-700',
    bgColor: 'bg-emerald-100',
    description: 'Professional writer assisting with application',
  },
  reviewer: {
    key: 'reviewer',
    label: 'Reviewer',
    color: 'text-amber-700',
    bgColor: 'bg-amber-100',
    description: 'Internal reviewer providing feedback',
  },
  admin: {
    key: 'admin',
    label: 'Administrator',
    color: 'text-slate-700',
    bgColor: 'bg-slate-100',
    description: 'Administrative support and oversight',
  },
};

// UI configuration for review status
export interface StatusConfig {
  key: ReviewStatus;
  label: string;
  color: string;
  bgColor: string;
  icon: string;
}

export const REVIEW_STATUS_CONFIGS: Record<ReviewStatus, StatusConfig> = {
  pending: {
    key: 'pending',
    label: 'Pending',
    color: 'text-gray-600',
    bgColor: 'bg-gray-100',
    icon: 'ClockIcon',
  },
  in_review: {
    key: 'in_review',
    label: 'In Review',
    color: 'text-blue-600',
    bgColor: 'bg-blue-100',
    icon: 'EyeIcon',
  },
  approved: {
    key: 'approved',
    label: 'Approved',
    color: 'text-green-600',
    bgColor: 'bg-green-100',
    icon: 'CheckCircleIcon',
  },
  rejected: {
    key: 'rejected',
    label: 'Rejected',
    color: 'text-red-600',
    bgColor: 'bg-red-100',
    icon: 'XCircleIcon',
  },
  escalated: {
    key: 'escalated',
    label: 'Escalated',
    color: 'text-orange-600',
    bgColor: 'bg-orange-100',
    icon: 'ExclamationTriangleIcon',
  },
};

// UI configuration for review actions
export interface ActionConfig {
  key: ReviewAction;
  label: string;
  color: string;
  bgColor: string;
  icon: string;
  description: string;
}

export const REVIEW_ACTION_CONFIGS: Record<ReviewAction, ActionConfig> = {
  approved: {
    key: 'approved',
    label: 'Approve',
    color: 'text-green-700',
    bgColor: 'bg-green-100',
    icon: 'CheckIcon',
    description: 'Approve and advance to next stage',
  },
  rejected: {
    key: 'rejected',
    label: 'Reject',
    color: 'text-red-700',
    bgColor: 'bg-red-100',
    icon: 'XMarkIcon',
    description: 'Reject the application at this stage',
  },
  returned: {
    key: 'returned',
    label: 'Return for Revision',
    color: 'text-amber-700',
    bgColor: 'bg-amber-100',
    icon: 'ArrowUturnLeftIcon',
    description: 'Return to previous stage for changes',
  },
  commented: {
    key: 'commented',
    label: 'Add Comment',
    color: 'text-blue-700',
    bgColor: 'bg-blue-100',
    icon: 'ChatBubbleLeftIcon',
    description: 'Add feedback without changing status',
  },
};
